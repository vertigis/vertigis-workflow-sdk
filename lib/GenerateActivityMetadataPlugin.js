// @ts-check
"use strict";

import * as path from "path";
import { Project } from "ts-morph";
import webpack from "webpack";
import { pathToFileURL } from "url";

import projPaths from "../config/paths.js";
import { getProjectMetadata } from "./compilerUtils.js";

/**
 * @type {string}
 */
const projUuid = (await import(pathToFileURL(path.join(projPaths.projRoot, "uuid.js")).href)).default;

const pluginName = "ActivityMetadataWebpackPlugin";

class GenerateActivityMetadataPlugin {
    apply(compiler) {
        compiler.hooks.thisCompilation.tap(pluginName, compilation => {
            compilation.hooks.processAssets.tap(
                {
                    name: pluginName,
                    stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
                },
                () => {
                    const project = new Project({
                        tsConfigFilePath: path.join(projPaths.projRoot, "tsconfig.json"),
                    });
                    const projectExportsFile = project.getSourceFileOrThrow("index.ts");
                    let manifest;
                    try {
                        manifest = getProjectMetadata(projectExportsFile, projUuid);
                    } catch (/** @type {any} */ err) {
                        // We rearrange the properties a bit here because webpack logs it as:
                        //   ERROR in ${err.message}
                        //   ${err.details}
                        err.details = err.message;
                        err.message = pluginName;
                        compilation.errors.push(err);
                    }

                    const manifestString = JSON.stringify(
                        manifest,
                        null,
                        process.env.NODE_ENV === "development" ? 2 : 0
                    );

                    compilation.emitAsset("activitypack.json", new webpack.sources.RawSource(manifestString));
                }
            );
        });
    }
}

export default GenerateActivityMetadataPlugin;
