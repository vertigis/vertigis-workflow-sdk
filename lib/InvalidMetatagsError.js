"use strict";

import webpack from "webpack";

class InvalidMetatagsError extends webpack.WebpackError {
    /** @param {string} message Error message */
    constructor(message) {
        super(message);
        this.name = "InvalidMetatagsError";
        this.hideStack = true;

        Error.captureStackTrace(this, this.constructor);
    }
}

export default InvalidMetatagsError;
