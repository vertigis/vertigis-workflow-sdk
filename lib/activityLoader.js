// @ts-check
"use strict";

import { createWrappedNode, ts } from "ts-morph";
import * as path from "path";
import { pathToFileURL } from "url";

import paths from "../config/paths.js";
import { isActivityDeclaration } from "./compilerUtils.js";

/**
 * Inject required members into activity class declarations.
 * @param {string} content The raw module source
 * @returns {Promise<string>}
 */
export default async function (content) {
    const sourceFile = ts.createSourceFile("activity.ts", content, ts.ScriptTarget.ES5, true);

    /**
     * @param {ts.ClassDeclaration} node
     * @param {string} source
     * @param {string} addition
     */
    function inject(node, source, addition) {
        // Inject just before the closing brace.
        const injectIndex = node.end - 1;
        return source.substring(0, injectIndex) + addition + source.substring(injectIndex);
    }

    for (const node of sourceFile.statements) {
        if (isActivity(node)) {
            if (!node.name) {
                throw new Error("Activity classes need to be named. `export default class ...` is not permitted.");
            }

            const needsAction = !node.members.find(member => member.name && member.name.getText() === "action");
            const needsSuite = !node.members.find(member => member.name && member.name.getText() === "suite");

            if (needsAction) {
                content = inject(
                    node,
                    content,
                    `\nstatic action = "uuid:${await getProjectUuid()}::${node.name.getText()}"\n`
                );
            }
            if (needsSuite) {
                content = inject(node, content, `\nstatic suite = "uuid:${await getProjectUuid()}"\n`);
            }
        }
    }

    return content;
}

/**
 * @returns {Promise<string>}
 */
async function getProjectUuid() {
    return (await import(pathToFileURL(path.join(paths.projRoot, "uuid.js")).href)).default;
}

/**
 * @param {ts.Node} node
 * @returns {node is ts.ClassDeclaration}
 */
function isActivity(node) {
    return isActivityDeclaration(createWrappedNode(node));
}
