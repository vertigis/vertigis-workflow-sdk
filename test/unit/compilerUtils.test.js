import { Project } from "ts-morph";
import { getProjectMetadata } from "../../lib/compilerUtils";
import InvalidMetatagsError from "../../lib/InvalidMetatagsError";

// Maintain a single instance of a project to avoid the large cost of
// initialization (~3s).
const project = new Project();
let sourceFiles = [];

// Clean up any source files created during the test run.
afterEach(() => {
    for (const sourceFile of sourceFiles) {
        project.removeSourceFile(sourceFile);
    }
    sourceFiles = [];
});

/**
 * @param {string} fileName
 * @param {string} sourceFileText
 */
function createSourceFile(fileName, sourceFileText) {
    const sourceFile = project.createSourceFile(fileName, sourceFileText, { overwrite: true });
    sourceFiles.push(sourceFile);
    return sourceFile;
}

const uuid = "proj-id-123";

describe("getProjectMetadata", () => {
    describe("activities", () => {
        it("passes basic sanity", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            interface TestActivityInputs {
                /**
                 * Input1Comment
                 * @clientOnly
                 * @defaultExpressionHint Input1DefaultExpressionHint
                 * @defaultValue Input1DefaultValue
                 * @deprecated Input1Deprecated
                 * @description Input1Description
                 * @displayName Input1DisplayName
                 * @helpUrl http://help/Activity/Input1
                 * @hidden
                 * @onlineOnly
                 * @placeholder Input1Placeholder
                 * @required
                 * @supportedApps gvh
                 * @tag Input1Tag1 value1
                 * @tag Input1Tag2 value2
                 */
                input1: string;

                input2: number;
            }

            interface TestActivityOutputs {
                /**
                 * Output1Comment
                 * @clientOnly
                 * @defaultExpressionHint Output1DefaultExpressionHint
                 * @defaultValue Output1DefaultValue
                 * @deprecated Output1Deprecated
                 * @description Output1Description
                 * @displayName Output1DisplayName
                 * @helpUrl http://help/Activity/Output1
                 * @hidden
                 * @onlineOnly
                 * @placeholder Output1Placeholder
                 * @required
                 * @supportedApps gxw
                 * @tag Output1Tag1 value1
                 * @tag Output1Tag2 value2
                 */
                output1: string;

                output2: number;
            }

            /**
             * ActivityComment
             * @category ActivityCategory
             * @clientOnly
             * @defaultName ActivityDefaultName
             * @deprecated ActivityDeprecated
             * @description ActivityDescription
             * @displayName ActivityDisplayName
             * @helpUrl http://help/TestActivity
             * @hidden
             * @onlineOnly
             * @supportedApps gvh, gxw
             * @tag ActivityTag1 value1
             * @tag ActivityTag2 value2
             * @tag ActivityTag3
             */
            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }

            /**
             * Should not be included as it doesn't implement 'IActivityHandler'
             */
            export class NotAnActivity {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities).toMatchSnapshot();
        });

        it("produces minimal defaults when JSDoc comments are absent", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            interface TestActivityInputs {
                inputA: string;
            }

            interface TestActivityOutputs {
                outputB: string;
            }

            export class TestActivity implements IActivityHandler {
                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities[0].category).toBe("Custom Activities");
            expect(activities[0].displayName).toBe("Test Activity");
            expect(activities[0].inputs.inputA.displayName).toBe("Input A");
            expect(activities[0].outputs.outputB.displayName).toBe("Output B");
        });

        it("rejects invalid metadata", () => {
            // Activity: Client & Server tags
            let activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            /**
             * @clientOnly
             * @serverOnly
             */
            export class TestActivity implements IActivityHandler {
                execute(inputs: any): any {
                    return undefined;
                }
            }
            `;
            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                InvalidMetatagsError
            );

            // Activity: Supported & Unsupported Apps
            activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            /**
             * @supportedApps gvh
             * @unsupportedApps wab
             */
            export class TestActivity implements IActivityHandler {
                execute(inputs: any): any {
                    return undefined;
                }
            }
            `;
            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                InvalidMetatagsError
            );

            // Input: Client & Server tags
            activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            interface TestActivityInputs {
                /**
                 * @clientOnly
                 * @serverOnly
                 */
                input1: number;
                input2: string;
            }
            export class TestActivity implements IActivityHandler {
                execute(inputs: TestActivityInputs): any {
                    return undefined;
                }
            }
            `;
            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                InvalidMetatagsError
            );

            // Input: Supported & Unsupported Apps
            activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            interface TestActivityInputs {
                /**
                 * @supportedApps gvh
                 * @unsupportedApps wab
                 */
                input1: number;
                input2: string;
            }
            export class TestActivity implements IActivityHandler {
                execute(inputs: TestActivityInputs): any {
                    return undefined;
                }
            }
            `;
            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                InvalidMetatagsError
            );

            // Output: Client & Server tags
            activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            interface TestActivityOutputs {
                /**
                 * @clientOnly
                 * @serverOnly
                 */
                output1: number;
                output2: string;
            }
            export class TestActivity implements IActivityHandler {
                execute(inputs: any): TestActivityOutputs {
                    return { output1: 12, output2: "fred" };
                }
            }
            `;
            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                InvalidMetatagsError
            );

            // Input: Supported & Unsupported Apps
            activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            interface TestActivityOutputs {
                /**
                 * @supportedApps gvh
                 * @unsupportedApps wab
                 */
                output1: number;
                output2: string;
            }
            export class TestActivity implements IActivityHandler {
                execute(inputs: any): TestActivityOutputs {
                    return { output1: 12, output2: "fred" };
                }
            }
            `;
            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                InvalidMetatagsError
            );
        });

        it("unwraps `Promise<T>` output type to `T`", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            interface TestActivityOutputs {
                output1: string;
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities[0].outputs.output1.typeName).toBe("string");
        });

        it("has no outputs when output type is `void`", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): void {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(Object.keys(activities[0].outputs)).toHaveLength(0);
        });

        it("has no outputs when output type is `Promise<void>`", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): Promise<void> {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(Object.keys(activities[0].outputs)).toHaveLength(0);
        });

        it("handles Promise<T> inputs/outputs", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            interface TestActivityInputs {
                input1: Promise<string>;
            }

            interface TestActivityOutputs {
                output1: Promise<boolean>;
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities[0].inputs.input1.typeName).toBe("Promise<string>");
            expect(activities[0].outputs.output1.typeName).toBe("Promise<boolean>");
        });

        it("handles complex types", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            interface TestActivityInputs {
                input1: {
                    foo: string;
                };
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities[0].inputs.input1.typeName).toBe(
                `{
    foo: string;
}`
            );
        });

        it("handles union types", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";

            interface TestActivityInputs {
                input1: "foo" | "bar" | string;
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities[0].inputs.input1.typeName).toBe(`"foo" | "bar" | string`);
        });

        it("handles unknown imported types", () => {
            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            import { MyInterface } from "./other";

            interface TestActivityInputs {
                input1: MyInterface;
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(activities[0].inputs.input1.typeName).toBe("MyInterface");
        });

        it("handles imported types", () => {
            const interfaceSource = `
            export interface MyInterface {
                foo: string;
            };
            `;
            sourceFiles.push(project.createSourceFile("MyInterface.ts", interfaceSource));

            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            import { MyInterface } from "../MyInterface";

            interface TestActivityInputs {
                input1: MyInterface;
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;
            sourceFiles.push(project.createSourceFile("activities/TestActivity.ts", activitySource));

            const indexSource = `
            export * from "./activities/TestActivity";
            `;

            const { activities } = getProjectMetadata(createSourceFile("index.ts", indexSource), uuid);

            expect(activities[0].inputs.input1.typeName).toBe("MyInterface");
        });

        it("handles absolute path sources", () => {
            // This test was specifically added to ensure type names do not contain inline imports
            // See https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
            const interfaceSource = `
            export interface MyInterface {
                foo: string;
            };
            `;
            sourceFiles.push(project.createSourceFile("c:/temp/MyInterface.ts", interfaceSource));

            const activitySource = `
            import { IActivityHandler } from "@vertigis/workflow/IActivityHandler";
            import { MyInterface } from "../MyInterface";

            interface TestActivityInputs {
                input1: MyInterface;
            }

            export class TestActivity implements IActivityHandler {
                static readonly action = "fake-action";
                static readonly suite = "fake-suite";

                execute(inputs: TestActivityInputs): TestActivityOutputs {
                    return {};
                }
            }
            `;
            sourceFiles.push(project.createSourceFile("c:/temp/activities/TestActivity.ts", activitySource));

            const indexSource = `
            export * from "./activities/TestActivity";
            `;

            const { activities } = getProjectMetadata(createSourceFile("c:/temp/index.ts", indexSource), uuid);

            expect(activities[0].inputs.input1.typeName).toBe("MyInterface");
        });
    });

    describe("elements", () => {
        it("passes basic sanity", () => {
            const activitySource = `
            import { FormElementProps, FormElementRegistration } from "@vertigis/workflow";
            import React from "react";

            interface Props extends FormElementProps<string> {
                /**
                 * Input1Comment
                 * @clientOnly
                 * @defaultExpressionHint Input1DefaultExpressionHint
                 * @defaultValue Input1DefaultValue
                 * @deprecated Input1Deprecated
                 * @description Input1Description
                 * @displayName Input1DisplayName
                 * @helpUrl http://help/Activity/Input1
                 * @hidden
                 * @onlineOnly
                 * @placeholder Input1Placeholder
                 * @required
                 * @unsupportedApps wab
                 */
                input1: string;

                input2: number;
            }

            /**
             * ElementComment
             * @category ElementCategory
             * @defaultName ElementDefaultName
             * @deprecated ElementDeprecated
             * @description ElementDescription
             * @displayName ElementDisplayName
             * @helpUrl http://help/TestElement
             * @unsupportedApps wab
             */
            const Foo = (props: Props) => null;

            export const fooRegistration: FormElementRegistration<Props> = {
                component: Foo,
                id: "foo",
            };

            /**
             * ElementComment
             * @category ElementCategory
             * @defaultName ElementDefaultName
             * @deprecated ElementDeprecated
             * @description ElementDescription
             * @displayName ElementDisplayName
             * @helpUrl http://help/TestElement
             * @unsupportedApps wab
             */
            function Bar(props: Props) { return null; };

            export const barRegistration: FormElementRegistration<Props> = {
                component: Bar,
                id: "bar",
            };

            /**
             * ElementComment
             * @category ElementCategory
             * @defaultName ElementDefaultName
             * @deprecated ElementDeprecated
             * @description ElementDescription
             * @displayName ElementDisplayName
             * @helpUrl http://help/TestElement
             * @unsupportedApps wab
             */
            class Baz extends React.Component<Props> {
                render() { return null; }
            };

            export const bazRegistration: FormElementRegistration<Props> = {
                component: Baz,
                id: "baz",
            };
            `;

            const { elements } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(elements).toMatchSnapshot();
        });

        it("includes elements with basic props interface", () => {
            const activitySource = `
            import { FormElementProps, FormElementRegistration } from "@vertigis/workflow";
            import React from "react";

            interface Props extends FormElementProps<string> {}

            const Foo = (props: Props) => null;

            export const fooRegistration: FormElementRegistration<Props> = {
                component: Foo,
                id: "foo",
            };

            const Bar = (props: FormElementProps) => null;

            export const barRegistration: FormElementRegistration = {
                component: Bar,
                id: "bar",
            };

            const Baz = () => null;

            export const bazRegistration: FormElementRegistration = {
                component: Baz,
                id: "baz",
            };
            `;

            const { elements } = getProjectMetadata(createSourceFile("index.ts", activitySource), uuid);

            expect(elements.map(element => element.inputs)).toEqual([{}, {}, {}]);
        });

        it("throws error for invalid `component` value", () => {
            const activitySource = `
            import { FormElementProps, FormElementRegistration } from "@vertigis/workflow";

            const Foo = 42;

            export const fooRegistration: FormElementRegistration = {
                component: Foo,
                id: "foo"
            };
            `;

            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                `"component" property of element registration isn't a function or class component.`
            );
        });

        it("throws error for shorthand `id` expression", () => {
            const activitySource = `
            import { FormElementProps, FormElementRegistration } from "@vertigis/workflow";

            const Foo = () => null;

            const id = "foo";

            export const fooRegistration: FormElementRegistration = {
                component: Foo,
                id
            };
            `;

            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                'Must directly assign value to "id" property of element registration'
            );
        });

        it("throws error for incorrect type of `id` value", () => {
            const activitySource = `
            import { FormElementProps, FormElementRegistration } from "@vertigis/workflow";

            const Foo = () => null;

            export const fooRegistration: FormElementRegistration = {
                component: Foo,
                id: 42
            };
            `;

            expect(() => getProjectMetadata(createSourceFile("index.ts", activitySource), uuid)).toThrow(
                `Value of "id" property of element registration must be a string literal.`
            );
        });
    });
});
