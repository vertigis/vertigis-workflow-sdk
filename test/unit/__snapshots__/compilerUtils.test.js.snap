// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProjectMetadata activities passes basic sanity 1`] = `
[
  {
    "action": "fake-action",
    "category": "ActivityCategory",
    "clientOnly": true,
    "defaultName": "ActivityDefaultName",
    "deprecated": "ActivityDeprecated",
    "description": "ActivityDescription",
    "displayName": "ActivityDisplayName",
    "helpUrl": "http://help/TestActivity",
    "inputs": {
      "input1": {
        "clientOnly": true,
        "defaultExpressionHint": "Input1DefaultExpressionHint",
        "defaultValue": "Input1DefaultValue",
        "deprecated": "Input1Deprecated",
        "description": "Input1Description",
        "displayName": "Input1DisplayName",
        "isHidden": true,
        "isRequired": true,
        "name": "input1",
        "noExpressions": undefined,
        "onlineOnly": true,
        "placeholder": "Input1Placeholder",
        "serverOnly": undefined,
        "supportedApps": [
          "gvh",
        ],
        "typeName": "string",
        "unsupportedApps": undefined,
      },
      "input2": {
        "clientOnly": undefined,
        "defaultExpressionHint": undefined,
        "defaultValue": undefined,
        "deprecated": undefined,
        "description": undefined,
        "displayName": "Input2",
        "isHidden": undefined,
        "isRequired": undefined,
        "name": "input2",
        "noExpressions": undefined,
        "onlineOnly": undefined,
        "placeholder": undefined,
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "number",
        "unsupportedApps": undefined,
      },
    },
    "isHidden": true,
    "onlineOnly": true,
    "outputs": {
      "output1": {
        "clientOnly": true,
        "defaultExpressionHint": "Output1DefaultExpressionHint",
        "defaultValue": "Output1DefaultValue",
        "deprecated": "Output1Deprecated",
        "description": "Output1Description",
        "displayName": "Output1DisplayName",
        "isHidden": true,
        "isRequired": true,
        "name": "output1",
        "noExpressions": undefined,
        "onlineOnly": true,
        "placeholder": "Output1Placeholder",
        "serverOnly": undefined,
        "supportedApps": [
          "gxw",
        ],
        "typeName": "string",
        "unsupportedApps": undefined,
      },
      "output2": {
        "clientOnly": undefined,
        "defaultExpressionHint": undefined,
        "defaultValue": undefined,
        "deprecated": undefined,
        "description": undefined,
        "displayName": "Output2",
        "isHidden": undefined,
        "isRequired": undefined,
        "name": "output2",
        "noExpressions": undefined,
        "onlineOnly": undefined,
        "placeholder": undefined,
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "number",
        "unsupportedApps": undefined,
      },
    },
    "serverOnly": undefined,
    "suite": "fake-suite",
    "supportedApps": [
      "gvh",
      "gxw",
    ],
    "unsupportedApps": undefined,
  },
]
`;

exports[`getProjectMetadata elements passes basic sanity 1`] = `
[
  {
    "deprecated": "ElementDeprecated",
    "description": "ElementDescription",
    "displayName": "ElementDisplayName",
    "helpUrl": "http://help/TestElement",
    "id": "foo",
    "inputs": {
      "input1": {
        "clientOnly": true,
        "defaultExpressionHint": "Input1DefaultExpressionHint",
        "defaultValue": "Input1DefaultValue",
        "deprecated": "Input1Deprecated",
        "description": "Input1Description",
        "displayName": "Input1DisplayName",
        "isHidden": true,
        "isRequired": true,
        "name": "input1",
        "noExpressions": undefined,
        "onlineOnly": true,
        "placeholder": "Input1Placeholder",
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "string",
        "unsupportedApps": [
          "wab",
        ],
      },
      "input2": {
        "clientOnly": undefined,
        "defaultExpressionHint": undefined,
        "defaultValue": undefined,
        "deprecated": undefined,
        "description": undefined,
        "displayName": "Input2",
        "isHidden": undefined,
        "isRequired": undefined,
        "name": "input2",
        "noExpressions": undefined,
        "onlineOnly": undefined,
        "placeholder": undefined,
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "number",
        "unsupportedApps": undefined,
      },
    },
    "suite": "uuid:proj-id-123",
    "supportedApps": undefined,
    "unsupportedApps": [
      "wab",
    ],
  },
  {
    "deprecated": "ElementDeprecated",
    "description": "ElementDescription",
    "displayName": "ElementDisplayName",
    "helpUrl": "http://help/TestElement",
    "id": "bar",
    "inputs": {
      "input1": {
        "clientOnly": true,
        "defaultExpressionHint": "Input1DefaultExpressionHint",
        "defaultValue": "Input1DefaultValue",
        "deprecated": "Input1Deprecated",
        "description": "Input1Description",
        "displayName": "Input1DisplayName",
        "isHidden": true,
        "isRequired": true,
        "name": "input1",
        "noExpressions": undefined,
        "onlineOnly": true,
        "placeholder": "Input1Placeholder",
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "string",
        "unsupportedApps": [
          "wab",
        ],
      },
      "input2": {
        "clientOnly": undefined,
        "defaultExpressionHint": undefined,
        "defaultValue": undefined,
        "deprecated": undefined,
        "description": undefined,
        "displayName": "Input2",
        "isHidden": undefined,
        "isRequired": undefined,
        "name": "input2",
        "noExpressions": undefined,
        "onlineOnly": undefined,
        "placeholder": undefined,
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "number",
        "unsupportedApps": undefined,
      },
    },
    "suite": "uuid:proj-id-123",
    "supportedApps": undefined,
    "unsupportedApps": [
      "wab",
    ],
  },
  {
    "deprecated": "ElementDeprecated",
    "description": "ElementDescription",
    "displayName": "ElementDisplayName",
    "helpUrl": "http://help/TestElement",
    "id": "baz",
    "inputs": {
      "input1": {
        "clientOnly": true,
        "defaultExpressionHint": "Input1DefaultExpressionHint",
        "defaultValue": "Input1DefaultValue",
        "deprecated": "Input1Deprecated",
        "description": "Input1Description",
        "displayName": "Input1DisplayName",
        "isHidden": true,
        "isRequired": true,
        "name": "input1",
        "noExpressions": undefined,
        "onlineOnly": true,
        "placeholder": "Input1Placeholder",
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "string",
        "unsupportedApps": [
          "wab",
        ],
      },
      "input2": {
        "clientOnly": undefined,
        "defaultExpressionHint": undefined,
        "defaultValue": undefined,
        "deprecated": undefined,
        "description": undefined,
        "displayName": "Input2",
        "isHidden": undefined,
        "isRequired": undefined,
        "name": "input2",
        "noExpressions": undefined,
        "onlineOnly": undefined,
        "placeholder": undefined,
        "serverOnly": undefined,
        "supportedApps": undefined,
        "typeName": "number",
        "unsupportedApps": undefined,
      },
    },
    "suite": "uuid:proj-id-123",
    "supportedApps": undefined,
    "unsupportedApps": [
      "wab",
    ],
  },
]
`;
